src/main/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposer.java

```diff
@@ -245,10 +255,171 @@ protected String getProtoRpcFullMethodName(Service protoService, Method protoMet
+          VariableExpr.builder()
+              .setVariable(pathTemplateVar)
+              .setIsDecl(true)
+              .setIsStatic(true)
+              .setIsFinal(true)
+              .setScope(ScopeNode.PRIVATE)
+              .build();
+      ValueExpr valueExpr =
+          ValueExpr.withValue(StringObjectValue.withValue(routingHeader.pattern()));
+      Expr pathTemplateExpr =
+          AssignmentExpr.builder()
+              .setVariableExpr(pathTemplateVarExpr)
+              .setValueExpr(
+                  MethodInvocationExpr.builder()
+                      .setStaticReferenceType(pathTemplateType)
+                      .setMethodName("create")
+                      .setArguments(valueExpr)
+                      .setReturnType(pathTemplateType)
+                      .build())
+              .build();
+      Statement pathTemplateClassVar = ExprStatement.withExpr(pathTemplateExpr);
+      classStatements.add(pathTemplateClassVar);
+      Expr valueOfExpr =
+          MethodInvocationExpr.builder()
+              .setStaticReferenceType(TypeNode.STRING)
+              .setMethodName("valueOf")
+              .setArguments(requestFieldGetterExpr)
+              .build();
+      MethodInvocationExpr addParamsMethodExpr =
+          MethodInvocationExpr.builder()
+              .setExprReferenceExpr(routingHeadersBuilderVarNonDeclExpr)
+              .setMethodName("add")
+              .setArguments(valueOfExpr, routingHeaderKeyExpr, routingHeaderPatternExpr)
+              .build();
+
+      IfStatement ifStatement =
+          IfStatement.builder()
+              .setConditionExpr(
+                  fieldValuesNotNullConditionExpr(
+                      requestVarExpr, routingHeader.getDescendantFieldNames()))
+              .setBody(ImmutableList.of(ExprStatement.withExpr(addParamsMethodExpr)))
+              .build();
+      bodyStatements.add(ifStatement);
+    }
+
+    return MethodInvocationExpr.builder()
+        .setExprReferenceExpr(routingHeadersBuilderVarNonDeclExpr)
+        .setMethodName("build")
+        .setReturnType(returnType)
+        .build();
+  }
+
+  private Expr fieldValuesNotNullConditionExpr(
+      VariableExpr requestVarExpr, List<String> fieldNames) {
+    MethodInvocationExpr.Builder requestFieldGetterExprBuilder =
+        MethodInvocationExpr.builder().setExprReferenceExpr(requestVarExpr);
+    Expr fieldValuesNotNullExpr = null;
+    for (int i = 0; i < fieldNames.size(); i++) {
+      String currFieldName = fieldNames.get(i);
+      String bindingFieldMethodName =
+          String.format("get%s", JavaStyle.toUpperCamelCase(currFieldName));
+      requestFieldGetterExprBuilder =
+          requestFieldGetterExprBuilder.setMethodName(bindingFieldMethodName);
+      // set return type of each method invocation to String just to pass the validation for
+      // RelationalOperationExpr that both side of relational operation needs to be a valid equality
```

Comment:
Can  you please elaborate on this? RelationalOperationExpr seems to be used to check for equality with a `null`, which does not have any type, so it seems like it should work with any type.

Reply:
if we want to check `if (request.getName() != null)`, we have to set the `returnType` of `request.getName` to a concrete type(which we don't have that info at this moment), otherwise it would fail the validation since it's expecting the the type of left hand side and right hand side of `RelationalOperationExpr` should be comparable(which makes sense, to prevent something like `if (5 != null)`). In our case, we know we are either checking an Object or a String, so setting the `returnType` to String just to pass validation.